doc/man/man3/Obstacle.3
.TH "Obstacle" 3 "Version 1.0" "Zaxxon" \" -*- nroff -*-
.ad l
.nh
.SH NAME
Obstacle \- Still obstacles that spawn across the levels\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fR#include <Obstacle\&.h>\fP
.PP
Inherits \fBEntity\fP\&.
.SS "Public Types"

.in +1c
.ti -1c
.RI "enum \fBObstacleType\fP { \fBGAS_CAN\fP = 1, \fBSATELLITE\fP, \fBGREY_CANNON\fP, \fBGREEN_CANNON\fP, \fBMISSILE_UP\fP, \fBPLANE\fP, \fBSPACE_FUEL\fP, \fBGREEN_CANNON_RIGHT\fP }"
.br
.in -1c
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBObstacle\fP (sf::Vector3f, sf::Texture *, int, int)"
.br
.RI "Create a turret obstacle with shooting and direction\&. "
.ti -1c
.RI "\fBObstacle\fP (sf::Vector3f, sf::Texture *, int)"
.br
.RI "Create a non-turret obstacle (Still obstacle)\&. "
.ti -1c
.RI "\fB~Obstacle\fP () override"
.br
.RI "Clean up memory related to the obstacle class\&. "
.ti -1c
.RI "std::vector< sf::Vector3f > \fBgetBulletLocations\fP ()"
.br
.RI "Get positions for all bullets of the obstacle (For turrets)\&. "
.ti -1c
.RI "void \fBupdate\fP (sf::RenderWindow &, int playerZ, float)"
.br
.RI "Run the logic for this obstacle\&. "
.ti -1c
.RI "void \fBupdate\fP (sf::RenderWindow &, float) override"
.br
.RI "Default \fBEntity::update\fP override\&. "
.ti -1c
.RI "bool \fBisPresent\fP ()"
.br
.RI "Check if this obstacle exists\&. "
.ti -1c
.RI "bool \fBisTurret\fP ()"
.br
.RI "Checks if obstacle is a turret\&. "
.ti -1c
.RI "void \fBbulletKill\fP (int)"
.br
.RI "Get rid of obstacles bullet by index (For turrets)\&. "
.ti -1c
.RI "\fBObstacleType\fP \fBgetType\fP ()"
.br
.RI "Get what type of obstacle this is\&. "
.ti -1c
.RI "int \fBgetScore\fP ()"
.br
.RI "Gets the score to add when the obstacle is destroyed\&. "
.ti -1c
.RI "void \fBkill\fP (\fBAnimation::Anim\fP \fBanimation\fP=\fBAnimation::CHARACTER_DEATH\fP) override"
.br
.RI "Death animation\&. "
.in -1c

Public Member Functions inherited from \fBEntity\fP
.in +1c
.ti -1c
.RI "\fBEntity\fP ()"
.br
.RI "Setup \fBEntity\fP class variables\&. "
.ti -1c
.RI "virtual \fB~Entity\fP ()=0"
.br
.RI "Clean up memory related to \fBEntity\fP class\&. "
.ti -1c
.RI "int \fBgetAnimationState\fP ()"
.br
.ti -1c
.RI "void \fBsetPos\fP (sf::Vector3f pos)"
.br
.ti -1c
.RI "sf::Vector3f \fBgetPos\fP ()"
.br
.ti -1c
.RI "sf::FloatRect \fBgetBounds\fP ()"
.br
.ti -1c
.RI "sf::Vector2f \fBgetSpritePos\fP ()"
.br
.RI "Get the position of the related sprite for this entity\&. "
.in -1c
.SS "Additional Inherited Members"


Protected Attributes inherited from \fBEntity\fP
.in +1c
.ti -1c
.RI "sf::Texture * \fBspriteSheet\fP"
.br
.ti -1c
.RI "sf::Sprite * \fBsprite\fP"
.br
.ti -1c
.RI "\fBAnimation\fP \fBanimations\fP"
.br
.ti -1c
.RI "bool \fBdead\fP = false"
.br
.ti -1c
.RI "sf::Text \fBdebugText\fP"
.br
.ti -1c
.RI "sf::Font \fBdebugFont\fP"
.br
.in -1c
.SH "Detailed Description"
.PP 
Still obstacles that spawn across the levels\&. 
.PP
Definition at line \fB14\fP of file \fBObstacle\&.h\fP\&.
.SH "Member Enumeration Documentation"
.PP 
.SS "enum \fBObstacle::ObstacleType\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIGAS_CAN \fP\fP
.TP
\fB\fISATELLITE \fP\fP
.TP
\fB\fIGREY_CANNON \fP\fP
.TP
\fB\fIGREEN_CANNON \fP\fP
.TP
\fB\fIMISSILE_UP \fP\fP
.TP
\fB\fIPLANE \fP\fP
.TP
\fB\fISPACE_FUEL \fP\fP
.TP
\fB\fIGREEN_CANNON_RIGHT \fP\fP
.PP
Definition at line \fB29\fP of file \fBObstacle\&.h\fP\&.
.SH "Constructor & Destructor Documentation"
.PP 
.SS "Obstacle::Obstacle (sf::Vector3f pos, sf::Texture * tex, int delay, int dir)"

.PP
Create a turret obstacle with shooting and direction\&. 
.PP
\fBParameters\fP
.RS 4
\fIpos\fP 
.br
\fItex\fP 
.br
\fIdelay\fP (type 5 only)Delay in frames before a rocket begins shooting\&.
.br
\fIdir\fP 
.RE
.PP

.PP
Definition at line \fB11\fP of file \fBObstacle\&.cpp\fP\&.
.SS "Obstacle::Obstacle (sf::Vector3f pos, sf::Texture * tex, int type)"

.PP
Create a non-turret obstacle (Still obstacle)\&. 
.PP
\fBParameters\fP
.RS 4
\fIpos\fP 
.br
\fItex\fP 
.br
\fItype\fP 
.RE
.PP

.PP
Definition at line \fB91\fP of file \fBObstacle\&.cpp\fP\&.
.SS "Obstacle::~Obstacle ()\fR [override]\fP"

.PP
Clean up memory related to the obstacle class\&. 
.PP
Definition at line \fB136\fP of file \fBObstacle\&.cpp\fP\&.
.SH "Member Function Documentation"
.PP 
.SS "void Obstacle::bulletKill (int bullet)"

.PP
Get rid of obstacles bullet by index (For turrets)\&. 
.PP
\fBParameters\fP
.RS 4
\fIbullet\fP 
.RE
.PP

.PP
Definition at line \fB337\fP of file \fBObstacle\&.cpp\fP\&.
.SS "std::vector< sf::Vector3f > Obstacle::getBulletLocations ()"

.PP
Get positions for all bullets of the obstacle (For turrets)\&. 
.PP
\fBReturns\fP
.RS 4
Vector of vectors of 3 floats
.RE
.PP

.PP
Definition at line \fB160\fP of file \fBObstacle\&.cpp\fP\&.
.SS "int Obstacle::getScore ()"

.PP
Gets the score to add when the obstacle is destroyed\&. 
.PP
\fBReturns\fP
.RS 4
score
.RE
.PP

.PP
Definition at line \fB359\fP of file \fBObstacle\&.cpp\fP\&.
.SS "\fBObstacle::ObstacleType\fP Obstacle::getType ()"

.PP
Get what type of obstacle this is\&. 
.PP
\fBReturns\fP
.RS 4
An int
.RE
.PP

.PP
Definition at line \fB349\fP of file \fBObstacle\&.cpp\fP\&.
.SS "bool Obstacle::isPresent ()"

.PP
Check if this obstacle exists\&. 
.PP
\fBReturns\fP
.RS 4
A boolean
.RE
.PP

.PP
Definition at line \fB315\fP of file \fBObstacle\&.cpp\fP\&.
.SS "bool Obstacle::isTurret ()"

.PP
Checks if obstacle is a turret\&. 
.PP
\fBReturns\fP
.RS 4
Boolean
.RE
.PP

.PP
Definition at line \fB326\fP of file \fBObstacle\&.cpp\fP\&.
.SS "void Obstacle::kill (\fBAnimation::Anim\fP animation = \fR\fBAnimation::CHARACTER_DEATH\fP\fP)\fR [override]\fP, \fR [virtual]\fP"

.PP
Death animation\&. 
.PP
\fBParameters\fP
.RS 4
\fIanimation\fP 
.RE
.PP

.PP
Reimplemented from \fBEntity\fP\&.
.PP
Definition at line \fB147\fP of file \fBObstacle\&.cpp\fP\&.
.SS "void Obstacle::update (sf::RenderWindow & window, float gameSpeed)\fR [override]\fP, \fR [virtual]\fP"

.PP
Default \fBEntity::update\fP override\&. 
.PP
Implements \fBEntity\fP\&.
.PP
Definition at line \fB305\fP of file \fBObstacle\&.cpp\fP\&.
.SS "void Obstacle::update (sf::RenderWindow & window, int playerZ, float gameSpeed)"

.PP
Run the logic for this obstacle\&. 
.PP
\fBParameters\fP
.RS 4
\fIwindow\fP 
.RE
.PP

.PP
Definition at line \fB170\fP of file \fBObstacle\&.cpp\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Zaxxon from the source code\&.


src/Obstacle/Obstacle.cpp
#include "Obstacle.h"


/// <summary>
/// Create a turret obstacle with shooting and direction.
/// </summary>
/// <param name="pos"></param>
/// <param name="tex"></param>
/// <param name="delay">(type 5 only)Delay in frames before a rocket begins shooting.</param>
/// <param name="dir"></param>
Obstacle::Obstacle(sf::Vector3f pos, sf::Texture* tex, int delay,
                   int dir) : Entity()
{
	total = (rand() % 250) + 75;
	this->scoreIndicator = rand() % 3 + 1;
	this->type = static_cast<ObstacleType>(dir + 3);

	setPos(pos);
	turret = true;
	direction = dir;

	spriteSheet = tex;

	switch (dir)
	{
	//grey turrets
	case 0:
		sprite->setTexture((*spriteSheet));
		sprite->setTextureRect(sf::IntRect(8, 112, 29, 19));

		sprite->setPosition(translateTo2d(pos));
		sprite->setOrigin(sprite->getGlobalBounds().width / 2,
			sprite->getGlobalBounds().height / 2);
		break;
	//green turrets
	case 1:
		sprite->setTexture((*spriteSheet));
		sprite->setTextureRect(sf::IntRect(48, 117, 30, 17));

		sprite->setPosition(translateTo2d(pos));
		sprite->setOrigin(sprite->getGlobalBounds().width / 2,
			sprite->getGlobalBounds().height / 2);
		break;
	//shooting up & red shooting up
	case 4:
	case 2:
		spriteSheet = tex;
		setPos(pos + sf::Vector3f(-14, 0, 0));
		sprite->setTexture((*spriteSheet));
		sprite->setTextureRect(sf::IntRect(72, 69, 32, 30));
		sprite->setPosition(translateTo2d(pos));
		sprite->setOrigin(sf::Vector2f(sprite->getGlobalBounds().width / 2,
			sprite->getGlobalBounds().height / 2));
		rocketZ = delay;

		rocketExplosionSprite.setTexture(*spriteSheet);
		rocketExplosionSprite.setTextureRect(sf::IntRect(119, 71, 38, 27));
		rocketExplosionSprite.setPosition(
			translateTo2d(pos) + sf::Vector2f(0, 0));
		rocketExplosionSprite.setOrigin(sf::Vector2f(
			sprite->getGlobalBounds().width / 2,
			sprite->getGlobalBounds().height / 2));
		break;
	//right green turrets
	case 3:
		type = GREEN_CANNON_RIGHT;
		sprite->setTexture((*spriteSheet));
		sprite->setTextureRect(sf::IntRect(320, 153, 33, 22));

		sprite->setPosition(translateTo2d(pos));
		sprite->setOrigin(sprite->getGlobalBounds().width / 2,
			sprite->getGlobalBounds().height / 2);
		break;
	}

	if (dir == 4)
	{
		redRocket = true;
		this->direction = 2;
		this->type = MISSILE_UP;
	}
}


/// <summary>
/// Create a non-turret obstacle (Still obstacle).
/// </summary>
/// <param name="pos"></param>
/// <param name="tex"></param>
/// <param name="type"></param>
Obstacle::Obstacle(sf::Vector3f pos, sf::Texture* tex, int type) : Entity()
{
	/*
	KEY
	1 = gas can
	2 = satellite
	3 = plane
	4 = blue space gas can
	*/

	this->type = static_cast<ObstacleType>(type);
	direction = -1;
	setPos(pos);
	turret = false;
	spriteSheet = tex;

	sprite->setTexture((*spriteSheet));

	switch (type)
	{
	case 1:
		sprite->setTextureRect(sf::IntRect(87, 106, 30, 30));
		break;
	case 2:
		sprite->setTextureRect(sf::IntRect(129, 109, 24, 28));
		break;
	case 3:
		this->type = PLANE;
		sprite->setTextureRect(sf::IntRect(92, 35, 29, 25));
		break;
	case 4:
		this->type = SPACE_FUEL;
		sprite->setTextureRect(sf::IntRect(8, 195, 26, 29));
		break;
	}

	sprite->setPosition(translateTo2d(pos));
	sprite->setOrigin(sprite->getGlobalBounds().width / 2,
	                  sprite->getGlobalBounds().height / 2);
}


/// <summary>
/// Clean up memory related to the obstacle class
/// </summary>
Obstacle::~Obstacle()
{
	const int bulletsSize = bullets.size();
	for (int i = 0; i < bulletsSize; i++)
		delete bullets[i];
}

/// <summary>
/// Death animation.
/// </summary>
/// <param name="animation"></param>
void Obstacle::kill(Animation::Anim animation)
{
	if (type == 5)
		animation = Animation::ALT_DEATH;
	animations.run(sprite, Animation::RESET);
	animations.run(sprite, animation, isTurret() ? scoreIndicator : 0);
};


/// <summary>
/// Get positions for all bullets of the obstacle (For turrets).
/// </summary>
/// <returns>Vector of vectors of 3 floats</returns>
std::vector<sf::Vector3f> Obstacle::getBulletLocations()
{
	return bulletPositions;
}


/// <summary>
/// Run the logic for this obstacle.
/// </summary>
/// <param name="window"></param>
void Obstacle::update(sf::RenderWindow& window, int playerZ, float gameSpeed)
{
	//Checks if obstacle is on screen
	if (!getWindowViewRect(window).intersects(sprite->getGlobalBounds())
		|| animations.getState() == 1)
	{
		onScreen = false;
		return;
	}

	onScreen = true;
	bool rocketExplosion = playerZ <= -rocketZ, rocketFiring = playerZ <= -
		     rocketZ - 50;

	//Shooting mechanics for bassic turrets
	if (turret == true && direction != 2)
	{
		//Creates bullets for grey cannons
		if (count % total == 0 && direction == 0 && animations.getState() == 0)
		{
			bullets.push_back(new ObstacleBullet(
				getPos(), spriteSheet, AbstractBullet::BulletType::zBullet));
			bulletPositions.push_back(getPos());

			total = (rand() % 250) + 75;
		}

		//Creates bullets for green cannons
		else if (count % total == 0 && (direction == 1 || direction == 3)
			&& animations.getState() == 0)
		{
			switch (direction)
			{
			case 1:
				bullets.push_back(new ObstacleBullet(
					getPos(), spriteSheet,
					AbstractBullet::BulletType::xBulletL));
				break;
			case 3:
				bullets.push_back(new ObstacleBullet(
					getPos(), spriteSheet,
					AbstractBullet::BulletType::xBulletR));
				break;
			default:
				break;
			}

			bulletPositions.push_back(getPos());

			total = (rand() % 250) + 75;
		}
	}
	//Controls the firing of up shooting missiles which are delayed from shooting afer entering the screen
	else if (direction == 2 && (animations.getState() == 0 || animations.
		getState() == 6 || animations.getState() == 7))
	{
		if (rocketFiring)
		{
			//Is the animation of launch still starting if so reset it
			if (animations.getState() == 3)
				animations.run(sprite, Animation::RESET);

			//is it below its explode height
			if (getPos().y > 70.f)
			{
				setPos(sf::Vector3f(getPos().x, getPos().y - 0.5f, getPos().z));
				sprite->setPosition(translateTo2d(getPos()));
				if (animations.getState() == 0 || animations.getState() == 7)
					animations.run(sprite, Animation::ROCKET_FLICKER,
					               redRocket);
			}
			else if (animations.getState() == 7)
				kill(Animation::ALT_DEATH);
		}
		else if (rocketExplosion &&
			rocketAnimation.getState() < 3
		)
			rocketAnimation.run(&rocketExplosionSprite, Animation::LAUNCH);
	}

	//Gives bullets their direction
	for (unsigned int i = 0; i < bullets.size(); i++)
	{
		bulletPositions.at(i) = bullets.at(i)->getPos();
		bullets.at(i)->update(window, gameSpeed);
	}

	//Offset position on death
	if (animations.getState() == 4 && !moved)
	{
		moved = true;
		if (isTurret())
			setPos(getPos() + sf::Vector3f(0, 0, 0));
		else
			setPos(getPos() + sf::Vector3f(0, 15, 0));
	}
	sprite->setPosition(translateTo2d(getPos()));

	if (type != MISSILE_UP && (type != SPACE_FUEL || getPos().x < 15))
		window.draw(*sprite);
	if (type == MISSILE_UP && rocketExplosion)
	{
		if (rocketFiring)
			window.draw(*sprite);
		if (rocketAnimation.getState() != 4)
			window.draw(rocketExplosionSprite);
	}

	count = (count + 1) % total;

	//Moves Blue Space Gas Cans
	if (type == SPACE_FUEL)
	{
		setPos(sf::Vector3f(getPos().x + 1.f, getPos().y - 0.6f,
			getPos().z));
		sprite->setPosition(translateTo2d(sf::Vector3f(getPos().x + 1.f,
			getPos().y - 0.6f, getPos().z)));
	}

#ifndef NDEBUG
	debugText.setString(std::to_string(static_cast<int>(getPos().x)) + " " +
		std::to_string(static_cast<int>(getPos().y)) + " " + std::to_string(
			static_cast<int>(getPos().z))
		+ " " + std::to_string(animations.getState())
	);
	debugText.setPosition(sprite->getPosition());
	if (drawDebugText)
		window.draw(debugText);
#endif // !NDEBUG
}


/// <summary>
/// Default Entity::update override.
/// </summary>
void Obstacle::update(sf::RenderWindow& window, float gameSpeed)
{
	update(window, 0, gameSpeed);
}


/// <summary>
/// Check if this obstacle exists.
/// </summary>
/// <returns>A boolean</returns>
bool Obstacle::isPresent()
{
	return animations.getState() != 1 && animations.getState() != 2
		&& animations.getState() != 4 && onScreen;
}


/// <summary>
/// Checks if obstacle is a turret.
/// </summary>
/// <returns>Boolean</returns>
bool Obstacle::isTurret()
{
	return getType() == 3 || getType() == 4
		|| getType() == 8;
}


/// <summary>
/// Get rid of obstacles bullet by index (For turrets).
/// </summary>
/// <param name="bullet"></param>
void Obstacle::bulletKill(int bullet)
{
	delete bullets[bullet];
	bullets.erase(bullets.begin() + (bullet));
	bulletPositions.erase(bulletPositions.begin() + bullet);
}


/// <summary>
/// Get what type of obstacle this is.
/// </summary>
/// <returns>An int</returns>
Obstacle::ObstacleType Obstacle::getType()
{
	return type;
}


/// <summary>
/// Gets the score to add when the obstacle is destroyed.
/// </summary>
/// <returns>score</returns>
int Obstacle::getScore()
{
	int score = 0;

	switch (type)
	{
	case 1:
		score += 300;
		break;
	case 2:
		score += 1000;
		break;
	case 5:
		score += 150;
		break;
	case 6:
		score += 100;
		break;
	case 7:
		score += 300;
		break;
	default:
		switch (this->scoreIndicator)
		{
		case 1:
			score += 200;
			break;
		case 2:
			score += 500;
			break;
		case 3:
			score += 1000;
		}
	}

	return score;
}